### PR TITLE
Feature/cr 2290 not car azure ensure function app is using a managed identity

### PR DESCRIPTION
**Pull request by lirshindalman**

### CODE REVIEW

File: cloudrail/knowledge/context/azure/resources/webapp/azure_function_app.py [hunk 1]
```diff
@@ -6,24 +7,37 @@
...
+
 class AzureFunctionApp(AzureResource):
     """
         Attributes:
             name: Function app resource name.
             app_service_config: App service configuration.
             client_cert_mode: The mode of the Function App's client certificates requirement for incoming requests.
-            https_only: Indicates if the Function App only be accessed via HTTPS
+            https_only: Indicates if the Function App only be accessed via HTTPS.
+            managed identity: if the function app is managed or not.
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources/webapp/azure_function_app.py:28]: If this field was a boolean, I think we would use it this way, but this is more a config block.
So, perhaps we should say something like:
"The managed identity service configuration of this function app, if exists"

File: cloudrail/knowledge/context/azure/resources/webapp/azure_function_app.py [hunk 2]
```diff
@@ -6,24 +7,37 @@
...
     """
 
     def __init__(self, name: str,
                  client_cert_mode: FieldMode,
-                 https_only: bool):
+                 https_only: bool, identity: Optional[Identity]):
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources/webapp/azure_function_app.py:33]: I think the new field should be in a new line

File: cloudrail/knowledge/context/azure/resources_builders/scanner/function_app_builder.py [hunk 3]
```diff
@@ -12,10 +12,23 @@ def get_file_name(self) -> str:
 
     def do_build(self, attributes: dict) -> AzureFunctionApp:
         if AzureAppServiceType.FUNCTION_APP.value in attributes['kind']:
+            identity = None
             client_cert_mode: FieldMode = FieldMode('Required')
             if attributes['properties']['clientCertMode']:
                 client_cert_mode = FieldMode(attributes['properties']['clientCertMode'])
+            if attributes.get('identity'):
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/azure/resources_builders/scanner/function_app_builder.py:19]: Perhaps we can a bit simplify the code with:
"if identity_data := attributes.get('identity')"
This way, you won't need to use "attributes['identity']" later in the code here.

File: tests/knowledge/context/azure/test_function_app.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 from cloudrail.knowledge.context.azure.resources.webapp.azure_function_app import AzureFunctionApp
 from cloudrail.knowledge.context.azure.resources.webapp.constants import FieldMode
 from tests.knowledge.context.azure_context_test import AzureContextTest
-from tests.knowledge.context.test_context_annotation import context
+from tests.knowledge.context.test_context_annotation import context, TestOptions
```
Comment (Imanuel-Mizrahi) [inline tests/knowledge/context/azure/test_function_app.py:5]: We should remove the testoptions import

File: cloudrail/knowledge/rules/azure/non_context_aware/function_app_using_managed_identity_rule.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+class FunctionAppUseManagedIdentityRule(AzureBaseRule):
+
+    def get_id(self) -> str:
+        return 'non_car_function_app_using_managed_identity'
+
+    def execute(self, env_context: AzureEnvironmentContext, parameters: Dict[ParameterType, any]) -> List[Issue]:
+        issues: List[Issue] = []
+        for func_app in env_context.function_apps:
+            if func_app.identity is None or func_app.identity.type not in ['SystemAssigned', 'UserAssigned']:
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/rules/azure/non_context_aware/function_app_using_managed_identity_rule.py:16]: Please remove the second condition, we just need to check if there is no "func_app.identity"

File: tests/knowledge/rules/azure/non_context_aware/function_app_using_managed_identity_rule/use_user_assigned/main.tf [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  storage_account_access_key = azurerm_storage_account.storacc.primary_access_key
+
+  identity {
+    type = "UserAssigned"
+    identity_ids = [ 
+      azurerm_user_assigned_identity.identity.id
+    ]
+  }
+
+  // depends_on = [
```
Comment (Imanuel-Mizrahi) [inline tests/knowledge/rules/azure/non_context_aware/function_app_using_managed_identity_rule/use_user_assigned/main.tf:57]: Please remove the comment out lines

File: cloudrail/knowledge/rules/azure/non_context_aware/abstract_web_app_using_managed_identity_rule.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        pass
+
+    def execute(self, env_context: AzureEnvironmentContext, parameters: Dict[ParameterType, any]) -> List[Issue]:
+        issues: List[Issue] = []
+        for web_app in self.get_web_app_resources(env_context):
+            if not web_app.identity:
+                issues.append(
+                    Issue(
+                        f'The {web_app.get_type()} `{web_app.get_friendly_name()}` '
+                        f'does not have managed identity enabled.', web_app, web_app))
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/rules/azure/non_context_aware/abstract_web_app_using_managed_identity_rule.py:35]: The evidence should be "does not have managed identity enabled" or maybe:
"does not have managed identity attached / configured"?

**Review by Imanuel-Mizrahi — APPROVED** [review]

**Review by idan-aviv — APPROVED** [review]

**Pull request merged**